# Introduction to NumPy

Now you have seen list, tuple and dictionaries (and we hinted at sets). However, these are not particularly useful when you want to do numerical calculations, or matrix operations. You could try to write a for loop that would iterate through each element, however Python is very slow with operations like that.

A better (more Pythonic) way is to use the package known as `numpy`. Numpy arrays are similar to lists and tuples etc. in that they hold multiple values. However, all elements in a numpy array must be of the same data type.

Disregarding that small setback, however, we now have a lot more functionality in numpy arrays. It is convenient to think of the behaviour of numpy arrays as similar to a row vector.

In [ ]:
import numpy as np # This is how you can import packages in python.

The above line imports the package numpy (which includes all the functions we will be using). The part `as np` renames it to `np` (just for this program, don't worry), which is slightly more convenient.

Let us assign a numpy array to a variable x.

In [ ]:
x=np.array([1,2,3,4]) # np.array converts any array-like object to a numpy array
#x=[1,2,3,4]
y=2*x
print(x,y)

[1 2 3 4] [2 4 6 8]


Try the above code with a normal list. Now, hopefully it is clear why `numpy` is powerful. But that is just the beginning!

In [ ]:
x = np.arange(0,100, 1) # This is similar to the range function from earlier, except that it returns a numpy array
print(x[[4,5,6,7,12,9,10]])

[ 4  5  6  7 12  9 10]


Trying a similar trick with a list instead: (this should give an error!)

In [ ]:
x = [1,2,3,4,5,6,7,8,9,10]
print(x[[0,1,2]])

So `numpy` also supports indexing with lists, instead of only integers or slices, whereas lists don't.

Similar to the `len` function, numpy also has several functions to describe an array. These are also available as methods, which can be used as follows:

In [ ]:
x=np.array([[1,2,3],[4,5,6],[7,8,9]])
print(x)

[[1 2 3]
 [4 5 6]
 [7 8 9]]


In [ ]:
print(x.ndim) # number of dimensions. This is a 2D array
print(x.shape) # Shape of array
print(x.size) # Total number of elements in array (product of elements in shape)

2
(3, 3)
9


Multiplication, like addition is element wise. If matrix multiplication is required, we can either convert it into a numpy matrix or use `np.dot()`

In [ ]:
x*x

array([[ 1,  4,  9],
       [16, 25, 36],
       [49, 64, 81]])

In [ ]:
np.matrix(x)*np.matrix(x)

matrix([[ 30,  36,  42],
        [ 66,  81,  96],
        [102, 126, 150]])

In [ ]:
np.dot(x,x)

array([[ 30,  36,  42],
       [ 66,  81,  96],
       [102, 126, 150]])

Comparisons on numpy arrays are easier now

In [ ]:
x>5

array([[False, False, False],
       [False, False,  True],
       [ True,  True,  True]])

We can also use conditional expressions like indices

In [ ]:
y=np.copy(x) # Simply typing y=x doesnot make a new array, it just adds a reference to the old one.
             # Changes in y will then be reflected in x
y[y>5]=0
y

array([[1, 2, 3],
       [4, 5, 0],
       [0, 0, 0]])

If we just want the indices where a certain condition is satisified, we can use `np.where()`. This is one of the more **important** functions, as we can use it to filter out data according to conditions.

In [ ]:
x=np.array([0,1,2,3,4,5,6,7,0.5,0.6])
indices=np.where(x<5)
x[indices]

array([0. , 1. , 2. , 3. , 4. , 0.5, 0.6])

We can also delete elements using `numpy.delete`.

Note the array slicing. x is 2 dimensional, and requires 2 indices. `x[:, 0]` returns `np.array[1,4,7]`. Play around with different combinations to become comfortable with using indices, as well as `np.where`.

Also note the argument of the function called `axis`. For a 2D array, `axis=0` is along the rows, and `axis=1` is along the columns. `np.where` assigns the index to be `(np.array([2]))`. Removing index 2 along the columns results in y. If you wanted to remove the row instead, then use `axis=0`.

In [ ]:
x=np.array([[1,2,3],[4,5,6],[7,8,9]])
index=np.where(x[:,0]>5)
y=np.delete(x,index,axis=1)
print(y)

[[1 2]
 [4 5]
 [7 8]]


**Make sure you understand the part about indexing, slicing, and using `np.where` and `np.delete`, and are comfortable using it. Try out different arrays that you define yourself, of different dimensions and shapes, and experiment!!**

Other useful functions in numpy like `numpy.arange`, `numpy.linspace` often come in handy. We have already seen `arange`

In [ ]:
np.arange(0,1,0.1) # start(inclusive):stop(exclusive):step

array([0. , 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9])

In [ ]:
np.linspace(0,1,5) #start:stop(inclusive):number of elements

array([0.  , 0.25, 0.5 , 0.75, 1.  ])

That is not even it. Several mathematical functions are also available:

In [ ]:
x = np.arange(-np.pi, np.pi, 0.1)
print(x)
y = np.sin(x)
x_abs = np.abs(x)

[-3.14159265 -3.04159265 -2.94159265 -2.84159265 -2.74159265 -2.64159265
 -2.54159265 -2.44159265 -2.34159265 -2.24159265 -2.14159265 -2.04159265
 -1.94159265 -1.84159265 -1.74159265 -1.64159265 -1.54159265 -1.44159265
 -1.34159265 -1.24159265 -1.14159265 -1.04159265 -0.94159265 -0.84159265
 -0.74159265 -0.64159265 -0.54159265 -0.44159265 -0.34159265 -0.24159265
 -0.14159265 -0.04159265  0.05840735  0.15840735  0.25840735  0.35840735
  0.45840735  0.55840735  0.65840735  0.75840735  0.85840735  0.95840735
  1.05840735  1.15840735  1.25840735  1.35840735  1.45840735  1.55840735
  1.65840735  1.75840735  1.85840735  1.95840735  2.05840735  2.15840735
  2.25840735  2.35840735  2.45840735  2.55840735  2.65840735  2.75840735
  2.85840735  2.95840735  3.05840735]


In [ ]:
y2 = np.exp(x)
print(x)

[-3.14159265 -3.04159265 -2.94159265 -2.84159265 -2.74159265 -2.64159265
 -2.54159265 -2.44159265 -2.34159265 -2.24159265 -2.14159265 -2.04159265
 -1.94159265 -1.84159265 -1.74159265 -1.64159265 -1.54159265 -1.44159265
 -1.34159265 -1.24159265 -1.14159265 -1.04159265 -0.94159265 -0.84159265
 -0.74159265 -0.64159265 -0.54159265 -0.44159265 -0.34159265 -0.24159265
 -0.14159265 -0.04159265  0.05840735  0.15840735  0.25840735  0.35840735
  0.45840735  0.55840735  0.65840735  0.75840735  0.85840735  0.95840735
  1.05840735  1.15840735  1.25840735  1.35840735  1.45840735  1.55840735
  1.65840735  1.75840735  1.85840735  1.95840735  2.05840735  2.15840735
  2.25840735  2.35840735  2.45840735  2.55840735  2.65840735  2.75840735
  2.85840735  2.95840735  3.05840735]


If you thought `numpy` has revealed all its cards, it hasn't yet! <br>
You can also find the mean, standard deviation, maximum, minimum of an array, sort an array, do matrix computations and so much more.

If you feel right now that NumPy is vast, remember we have barely scratched the surface. In this module and the next is where your Googling skills will be put to the test.

If you want to find the official docstrings for any function, an easy way to get it from Jupyter is to use `?`. For example

In [ ]:
# Note that this is a very useful function, since we can use it to filter out data.
np.where?

As a final note, we will end the tutorial part of this notebook by introducing `np.loadtxt`, a function that can be used to load a txt file into a numpy array. (If you remember how difficult it was to import a text file using `open` and `.split()`, then you will appreciate this one!)

In [ ]:
beehive_data = np.loadtxt('Beehive_data.csv', delimiter=',')

Yes, that's it.

### Your assignment ...
... should you choose to accept it, will be the following:
1. Re-do Tutorial 1 assignment with numpy. (This should be fairly easy, but use only numpy functions to reproduce your results)
2. Use the data from `Beehive_data.csv` to find the distance to the Beehive Cluster. Read on ahead to find out more about the concepts in astronomy you will need to solve this

Data downloaded from [VizieR](https://cdsarc.unistra.fr/viz-bin/cat/V/19), from data used in [Piskunov 1980](https://ui.adsabs.harvard.edu/?#abs/1980BICDS..19...67P)

# Magnitudes in Astronomy

Magnitudes are a way to describe how bright an object (in our case, a star) is. It is similar to the decibel system for sound in that magnitudes are logarithmic. They can be calculated according to the formula
$$\rm m = -2.5 \log\left(\frac{F}{F_0}\right)$$
where F is the flux from the star (measured in W m$^{-2}$), and F$_0$ is a reference flux.

To read up more about magnitudes, hit up the [Wikipedia article on it](https://en.wikipedia.org/wiki/Magnitude_(astronomy)#History)

Now, we can calculate the flux of a star at some distance d away as
$$ F = \frac{L}{4\pi d^2} $$
where L is the Luminosity of the star (measured in W).

There are a lot of details we have skimmed over (for example, the use of filters) which we will visit again in more detail in `Tutorial 8: Image Reduction`.

As a final note on the data you have been given in `Beehive_data.csv`, the columns are respectively the apparent magnitude of the stars as seen from Earth, the logarithm of the Luminosity(in units of Luminosity of the Sun) of the stars (i.e. $\log(L/L_\odot)$, where $L_\odot$ is the luminosity of the sun), and the probability that the star belongs to the Beehive Cluster.

You must find the distance to the cluster (you are given that the absolute magnitude of the Sun is +4.83). You can do this in two ways:
1. Exclude all the stars with low probability of belonging to the cluster, and then calculate distance for each star, and find the mean.
2. Find the distance for all stars (including the low probability ones) and find the weighted average of the distances, where the weights are the probability.

Good Luck!